mapframe.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2018 Iguazio Systems Ltd.
+
+Licensed under the Apache License, Version 2.0 (the "License") with
+an addition restriction as set forth herein. You may not use this
+file except in compliance with the License. You may obtain a copy of
+the License at http://www.apache.org/licenses/LICENSE-2.0.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+implied. See the License for the specific language governing
+permissions and limitations under the License.
+
+In addition, you may not use the software for any purposes that are
+illegal under applicable law, and the grant of the foregoing license
+under the Apache 2.0 license is conditioned upon your compliance with
+such restriction.
+*/
+
+package frames
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+)
+
+// MapFrame is a frame based on map
+type MapFrame struct {
+	byIndex map[int]Column
+}
+
+// NewMapFrame returns a new MapFrame
+func NewMapFrame(columns []Column) (*MapFrame, error) {
+	if err := checkEqualLen(columns); err != nil {
+		return nil, err
+	}
+
+	byIndex := make(map[int]Column)
+	for i, col := range columns {
+		byIndex[i] = col
+	}
+
+	return &MapFrame{byIndex}, nil
+}
+
+// Columns returns the column names
+func (mf *MapFrame) Columns() []string {
+	names := make([]string, len(mf.byIndex))
+
+	for i := 0; i < len(mf.byIndex); i++ {
+		names[i] = mf.byIndex[i].Name() // TODO: Check if exists?
+	}
+
+	return names
+}
+
+// Len is the number of rows
+func (mf *MapFrame) Len() int {
+	for _, col := range mf.byIndex {
+		return col.Len()
+	}
+
+	return 0
+}
+
+// Column gets a column by name
+func (mf *MapFrame) Column(name string) (Column, error) {
+	// TODO: We can speed it up by calculating once, but then we'll use more memory
+	for _, col := range mf.byIndex {
+		if col.Name() == name {
+			return col, nil
+		}
+	}
+
+	return nil, fmt.Errorf("column %q not found", name)
+}
+
+// Slice return a new Frame with is slice of the original
+func (mf *MapFrame) Slice(start int, end int) (Frame, error) {
+	if err := validateSlice(start, end, mf.Len()); err != nil {
+		return nil, err
+	}
+
+	frameSlice := make([]Column, len(mf.byIndex))
+	for i, col := range mf.byIndex {
+		slice, err := col.Slice(start, end)
+		if err != nil {
+			return nil, errors.Wrapf(err, "can't get slice from %q", col.Name())
+		}
+
+		frameSlice[i] = slice
+	}
+
+	return NewMapFrame(frameSlice)
+}
+
+// MapFrameMessage is over-the-wire frame data
+type MapFrameMessage struct {
+	Columns   []string                       `msgpack:"columns"`
+	SliceCols map[string]*SliceColumnMessage `msgpack:"slice_cols,omitempty"`
+	LabelCols map[string]*LabelColumnMessage `msgpack:"label_cols,omitempty"`
+}
+
```

Comment:
@tebeka looks like some redundancy : 
1. Col names are in the col object 
2. Slice & Label can be part of the same list just different data types (Interface) 

also we need an indication which col is an index col (optional)